File: nasdaq-taq/src/main/java/org/jvirtanen/nasdaq/taq/TAQ.java
package org.jvirtanen.nasdaq.taq;

import static org.jvirtanen.util.Applications.*;

import com.paritytrading.foundation.ASCII;
import com.paritytrading.juncture.nasdaq.itch50.ITCH50Parser;
import com.paritytrading.nassau.binaryfile.BinaryFILEReader;
import com.paritytrading.nassau.binaryfile.BinaryFILEStatusListener;
import com.paritytrading.nassau.binaryfile.BinaryFILEStatusParser;
import com.paritytrading.parity.top.Market;
import java.io.File;
import java.io.IOException;

class TAQ {

    public static void main(String[] args) {
        if (args.length != 3)
            usage("nasdaq-taq <filename> <date> <instrument>");

        try {
            main(new File(args[0]), args[1], args[2]);
        } catch (IOException e) {
            error(e);
        }
    }

    private static void main(File file, String date, String instrument) throws IOException {
        TAQSink sink = new TAQSink(date, instrument);

        Market market = new Market(sink);

        ITCH50Parser listener = new ITCH50Parser(new ITCH50Source(market, ASCII.packLong(instrument), sink));

        BinaryFILEStatusListener statusListener = new BinaryFILEStatusListener() {

            @Override
            public void endOfSession() {
            }

        };

        BinaryFILEStatusParser statusParser = new BinaryFILEStatusParser(listener, statusListener);

        BinaryFILEReader reader = BinaryFILEReader.open(file, statusParser);

        market.open(ASCII.packLong(instrument));

        while (reader.read());

        sink.flush();
    }

}

File: nasdaq-taq/src/main/kotlin/org/jvirtanen/nasdaq/taq/TAQ.kt
package org.jvirtanen.nasdaq.taq

import org.jvirtanen.util.Applications.*

import com.paritytrading.foundation.ASCII
import com.paritytrading.juncture.nasdaq.itch50.ITCH50Parser
import com.paritytrading.nassau.util.BinaryFILE
import com.paritytrading.parity.top.Market
import java.io.File
import java.io.IOException

fun main(args: Array<String>) {
    if (args.size != 3)
        usage("nasdaq-taq <filename> <date> <instrument>")

    try {
        main(File(args[0]), args[1], args[2])
    } catch (e: IOException) {
        error(e)
    }
}

private fun main(file: File, date: String, instrument: String) {
    val sink   = TAQSink(date, instrument)
    val market = Market(sink)

    val listener = ITCH50Parser(ITCH50Source(market, ASCII.packLong(instrument), sink))

    market.open(ASCII.packLong(instrument))

    BinaryFILE.read(file, listener)

    sink.flush()
}
